htlcswitch/link_test.go

```diff
@@ -721,93 +726,186 @@ func TestChannelLinkCancelFullCommitment(t *testing.T) {
+	)
 
 	// In order to exercise this case, we'll now _manually_ modify the
-	// per-hop payload for amount to be the incorrect value.  The proper
-	// value of the amount to forward should be the amount that the
-	// receiving node expects to receive.
-	hops[0].FwdInfo.AmountToForward = 1
+	// per-hop payload for outgoing time lock to be the incorrect value.
+	// The proper value of the outgoing CLTV should be the policy set by
+	// the receiving node, instead we set it to be a value greater than the
+	// incoming HTLC timelock.
+	hops[0].FwdInfo.OutgoingCTLV = htlcExpiry + 1
 	firstHop := n.firstBobChannelLink.ShortChanID()
 	_, err = makePayment(
 		n.aliceServer, n.bobServer, firstHop, hops, amount, htlcAmt,
 		htlcExpiry,
 	).Wait(30 * time.Second)
-	if err == nil {
-		t.Fatalf("payment should have failed but didn't")
+	require.NotNil(t, err, "payment should have failed but didn't")
+
+	var rtErr ClearTextError
+	ok := errors.As(err, &rtErr)
+	require.True(t, ok)
+	require.IsType(
+		t, rtErr, err,
+		"expected a ClearTextError, instead got: %T", err,
+	)
+
+	switch rtErr.WireMessage().(type) {
+	case *lnwire.FailFinalIncorrectCltvExpiry:
+	default:
+		t.Fatalf("incorrect error, expected incorrect cltv expiry, "+
+			"instead have: %v", err)
+	}
+}
+
+// TestExitNodeHTLCUnderpaysPayloadAmount tests that when an exit node receives
+// an incoming HTLC, if the amount offered in the HTLC is less than the amount
+// encoded in the onion payload then the HTLC will be rejected with the
+// appropriate error.
+func TestExitNodeHTLCUnderpaysPayloadAmount(t *testing.T) {
+	t.Parallel()
+
+	f := func(underpayment_rand uint64) bool {
+		underpayment := lnwire.MilliSatoshi(
+			underpayment_rand%(btcutil.SatoshiPerBitcoin-1) + 1,
+		)
+		channels, _, err := createClusterChannels(
+			t, btcutil.SatoshiPerBitcoin*5,
+			btcutil.SatoshiPerBitcoin*5,
+		)
+		require.NoError(t, err, "unable to create channel")
+
+		n := newThreeHopNetwork(
+			t, channels.aliceToBob, channels.bobToAlice,
+			channels.bobToCarol, channels.carolToBob,
+			testStartingHeight,
+		)
+		require.NoError(t, n.start())
+		t.Cleanup(n.stop)
+
+		const amount = btcutil.SatoshiPerBitcoin
+		htlcAmt, htlcExpiry, hops := generateHops(
+			amount, testStartingHeight, n.firstBobChannelLink,
+		)
+
+		// In order to exercise this case, we'll now _manually_ modify
+		// the per-hop payload for amount to be the incorrect value.
+		// The acceptable values of the amount to forward should be less
+		// than the incoming HTLC value.
+		hops[0].FwdInfo.AmountToForward = amount + underpayment
+		firstHop := n.firstBobChannelLink.ShortChanID()
+		_, err = makePayment(
+			n.aliceServer, n.bobServer, firstHop, hops, amount,
+			htlcAmt, htlcExpiry,
+		).Wait(30 * time.Second)
+		assertFailureCode(t, err, lnwire.CodeFinalIncorrectHtlcAmount)
+
+		return err != nil
 	}
-	assertFailureCode(t, err, lnwire.CodeFinalIncorrectHtlcAmount)
+	err := quick.Check(f, &quick.Config{MaxCount: 20})
```

Comment:
IMO we should avoid using randomness in our unit tests, plus this randomness doesn't provide more utility here. If we wanna test random inputs we should do it in fuzz testing.

PS, looks like the package `quick` is frozen, so it won't be maintained?

Reply:
I very strongly disagree that testing should not be done with random inputs. Fixed tests do very little to test anything and secondly, by writing tests against random inputs we are forced to test and therefore document the true invariants that are supposed to hold in our code rather than asking the reader to guess at them by trying to do pattern reconstruction.

That said, idk if there's a functional difference here between what I want and what you're calling "fuzz testing" but I generally don't see the value in tests that do not have this setup. There are exceptions of course, but in my experience these are the only types of tests that have helped catch bugs.

Reply:
> Could you elaborate on this part? Coming from TDD background I think it makes little sense.

This will be a long one so bear with me :)

The main purpose of testing is to ensure that your code honors a *contract*. It's purpose is *not* to ensure a particular implementation of that contract. When we design code we expect it to have certain behaviors and it is that behavior we are trying to preserve. The issue with fixed-value unit tests is twofold. First, they only ensure that a singular point in the input space is valid. Secondly, because they allow the test writer to make certain assumptions about things, the test writer will naturally overlook or underdefine the overall contract that the code under test is supposed to honor.

The greatest assurance we can provide that our code is correct would be a proof. However, golang lacks a type system capable of providing proofs of correctness, and even with the languages that afford you this ability, you may still prefer not to write those types because the benefits you get don't outweigh the complexity cost of advanced type system features.

The least assurance we get is from selecting specific test cases: we can only be sure that the case that is tested is correct. There is a wide space of implementations that would satisfy the test case and not the *contract*. But the issue here is that the singular unit test cases are fundamentally incapable of describing a contract because they use a different quantifier.

A unit test case is an assurance claim using a `∃ (there exists)`. i.e. "there exists a value for which this assertion holds". A property test (what you call "fuzz testing") on the other hand attempts to approximate the `∀ (for all)`. i.e. "this assertion holds for all inputs". The claims being tested between these two quantifiers are *fundamentally different*. In fact, they are actually logical negations of one another. It is extremely rare where the genuine contract that your code has claimed to uphold is one where the quantifer is `∃`. In rare cases an individual unit test on a boundary condition (finite integer spaces, zero values, or other "special values" in your algebra) can be very helpful.

Property testing represents a point on this spectrum that gets you most of the way to a proof without having to go through the same rigor you would need to in order to get the same assurances out of a type system. It is for this reason that I believe that property testing should make up 90+% of your test suite. Any other testing strategy gives you comparatively little assurance that anything actually works, and does very little to help you understand how your codebase is *supposed to work*.

So maybe by now you're thinking that all of this is an argument for "just do both". But here's the thing, unit test cases are naturally subsumed by property tests, since the moment you sample your input values from your generation sources, it tests that individual case. So we lose nothing by omitting the unit test case entirely.

Writing code with high assurance, which as I understand from @saubyk is a huge priority to get LND to the point where it is being treated as mission critical software, is about making the strongest assertions you possibly can about the code and unit tests don't offer any benefit over property tests in this regard.

> Back to this specific test case, since we always have hops[0].FwdInfo.AmountToForward = amount + underpayment, how does the randomness help unless you are testing an overflow?

So if we examine this particular case with this perspective, the benefit afforded here is that we are demonstrating that the HTLC is rejected independent of *how much* the Payload is underpaid. This both tests a *stronger assertion* than a unit test ever could, *and* it documents the *actual property* we are looking for out of the code.

Reply:
My only concern with randomness in unit tests is reproducibility.  Better to have a consistently failing test in CI than one that might pass on a rerun (and thus hide the bug as "just a flaky test").  @ProofOfKeags Does `quick` seed the PRNG by default?

Besides that, I'm very interested in more property/fuzz testing of LND.

FWIW, this particular usage of `quick` could be trivially rewritten as a Go fuzz test:

```go
func FuzzExitNodeHTLCUnderpaysPayloadAmount(f *testing.F) {
  f.Fuzz(func(underpayment_rand uint64) {
    ...
  })
}
```

Reply:
> Does quick seed the PRNG by default?

Yes. In other language ecosystems I typically see it seeded with the git commit hash. However, I just looked it up and the default implementation of the PRNG seed is to use [the time](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/testing/quick/quick.go;l=198-204)

> My only concern with randomness in unit tests is reproducibility

I agree that repeated runs in the same pipeline should result in the same test failures. However, we actually *do* want the variation so that the pipeline can be capable of uncovering nonobvious edge cases in the long run.

Reply:
> I agree that repeated runs in the same pipeline should result in the same test failures. However, we actually _do_ want the variation so that the pipeline can be capable of uncovering nonobvious edge cases in the long run.

Our current approach for fuzz tests is to generate an input corpus from many CPU hours of fuzzing and submit them to the `lnd-fuzz` repo.  The LND CI then runs the fuzz tests on their seeds from `lnd-fuzz`.  This gives us determinism for CI and randomness when fuzzing.

One thing this setup is missing is a continuous fuzzing system to keep trying new inputs over time.  Something like https://github.com/lightningnetwork/lnd/issues/5965.

Reply:
Fwiw property testing and fuzz testing [serve different purposes](https://www.tedinski.com/2018/12/11/fuzzing-and-property-testing.html)

I'm not sure what I'm trying to accomplish here is in line with the traditional goals of fuzzing.

I'm trying to introduce (slowly) some algebraic property guarantees for the code that is under test, not necessarily trying to force UB states explicitly. There is significant overlap with these goals but I do want to make clear my aim and why I'm pushing that direction.

Reply:
Thanks for the reply and glad we are having this discussion!

> It is for this reason that I believe that property testing should make up 90+% of your test suite.

Ideally yes, in practice no. That'd be a nice goal, but we have limited engineering time. I think the key to making good property testing is to define properties and exhaust them all. Take this specific test case,

1. 20 out of infinite provides very little coverage but too much confidence. We'd think it's protected by property tests while it's not.
2. the state change happens in the test, not in the state machine. In other words, we are testing our test code, not the state machine. This is because we manually change `hops[0].FwdInfo.AmountToForward = amount + underpayment` before testing the state machine, so we are actually testing the addition works as expected, in other words, we are testing a behavior that's not made from the testing object.
3. not every test case needs to be property testing, aside from engineering time, functions like arithmetic operations should not rely on a random behavior but dedicated edge case tests are needed, like zero values, overflow, etc.

Since property test increases the testing time significantly, I don't think it's suitable for every package atm. Maybe we can start with the lower level packages like `lnwallet`.

> Fwiw property testing and fuzz testing serve different purposes

A good read, so is [this one](https://hypothesis.works/articles/what-is-property-based-testing/). Based on the author, the difference comes from where the random data is generated(user or machine).

> Writing code with high assurance

For sure. I think it's nice to have property tests, while I think the more urgent issue is to make our code testable - this means more interfaces, more defined responsibilities for each subsystem, and simply a higher coverage rate.

> However, we actually do want the variation so that the pipeline can be capable of uncovering nonobvious edge cases in the long run.

I think we need to seed it so it's deterministic. Instead of relying on repeated builds to generate variation, we should increase the input size in the test. Otherwise, here's a simple question - if a developer sees the property test fails in his/her PR, who's gonna fix it? This is already the question we are facing today in our builds, which will soon be fixed.

Reply:
Just to be clear, I'm not saying we should abandon the practice it here, but wanna have more discussion. Fwiw we already have similar tests in [watchtower](https://github.com/lightningnetwork/lnd/blob/master/watchtower/wtdb/codec_test.go#L128), [autopilot](https://github.com/lightningnetwork/lnd/blob/master/autopilot/choice_test.go#L62) and three other packages.